src/NUnitFramework/framework/Internal/TestNameGenerator.cs

```diff
@@ -0,0 +1,457 @@
+                    case "{7}":
+                    case "{8}":
+                    case "{9}":
+                        int index = token[1] - '0';
+                        _fragments.Add(new ArgumentFragment(index, 40));
+                        break;
+                    default:
+                        char c = token[1];
+                        if (token.Length >= 5 && token[2] == ':' && (c == 'a' || char.IsDigit(c)))
+                        {
+                            int length;
+
+                            // NOTE: The code would be much simpler using TryParse. However,
+                            // that method doesn't exist in the Compact Framework.
+                            try
+                            {
+                                length = int.Parse(token.Substring(3, token.Length - 4));
+                            }
+                            catch
+                            {
+                                length = -1;
+                            }
+                            if (length > 0)
+                            {
+                                if (c == 'a')
+                                    _fragments.Add(new ArgListFragment(length));
+                                else // It's a digit
+                                    _fragments.Add(new ArgumentFragment(c - '0', length));
+                                break;
+                            }
+                        }
+
+                        // Output the erroneous token to aid user in debugging
+                        _fragments.Add(new FixedTextFragment(token));
+                        break;
+                }
+
+                start = rcurly + 1;
+            }
+
+
+            // Output any trailing plain text
+            if (start < pattern.Length)
+                _fragments.Add(new FixedTextFragment(pattern.Substring(start)));
+        }
+
+        /// <summary>
+        /// Get the display name for a method
+        /// </summary>
+        /// <param name="method">A MethodInfo</param>
+        /// <returns>The display name</returns>
+        public string GetDisplayName(MethodInfo method)
+        {
+            return GetDisplayName(method, null);
+        }
+
+        /// <summary>
+        /// Get the display name for a method with args
+        /// </summary>
+        /// <param name="method">A MethodInfo</param>
+        /// <param name="args">Argument list for the method</param>
+        /// <returns>The display name</returns>
+        public string GetDisplayName(MethodInfo method, object[] args)
+        {
+            var result = new StringBuilder();
+
+            foreach (var fragment in _fragments)
+                result.Append(fragment.GetText(method, args));
+
+            return result.ToString();
+        }
+
+        #region Nested Classes Representing Name Fragments
+
+        private abstract class NameFragment
+        {
+            private const string THREE_DOTS = "...";
```

Comment:
I'm pretty sure all our devices will be happy enough. I'd like to be sure it displays correctly too, so I'll give it a try.

We currently use three dots in failure messages as well, but that's best dealt with as a separate issue.